### DISCUSSION TITLE
Export cue files to json using the Go API

### DISCUSSION THREAD

**Discussion by uris77** [cue-lang/cue#1188] [Q&A] [answered]
I'm trying to write a terraform provider for cue that will generate a json file from some cue files. So, I need to use Go to do this. I've been trying to use the `cmd` package to do this, but do this, but it ouputs the json file to stdout and I could not figure out how to redirect it to a file.

```
       command, err := cmd.New(args)
	if err != nil {
		panic(fmt.Errorf("failed to execute cmd: %v", err))
	}

	ctx := context.Background()

	err = command.Run(ctx)
	if err != nil {
		panic(fmt.Errorf("failed to run command %w", err))
	}
```

If the `cmd` package is the wrong approach, what functions in the `api` should I be focusing on?

**Comment by uris77** [comment, marked as answer]
I found a possible solution.  We can assign a `writer` to the `command`:

```
       command, err := cmd.New(args)
	if err != nil {
		panic(fmt.Errorf("failed to execute cmd: %v", err))
	}

	ctx := context.Background()
	file, err := os.Create("./temp.json")
	if err != nil {
		log.Fatal(err)
	}
	writer := bufio.NewWriter(file)
	command.SetOut(writer)
	//c, err := command.Command.ExecuteC()
	err = command.Run(ctx)
	if err != nil {
		panic(fmt.Errorf("failed to run command %w", err))
	}
```

**Comment by myitcv** [comment]
If you have a CUE value (which I assume you have if you have CUE files that you have loaded) you can simply marshal them to JSON:

```
go mod tidy
go run main.go

-- cue.mod/module.cue --
module: "mod.com"
-- go.mod --
module mod.com

go 1.16

require cuelang.org/go v0.4.0
-- main.cue --
package main

x: 5
-- main.go --
package main

import (
	"fmt"
	"log"

	"cuelang.org/go/cue/cuecontext"
	"cuelang.org/go/cue/load"
)

func main() {
	ctx := cuecontext.New()
	bps := load.Instances([]string{"."}, nil)
	v := ctx.BuildInstance(bps[0])
	byts, err := v.MarshalJSON()
	if err != nil {
		log.Fatal(err)
	}
	fmt.Printf("%s\n", byts)
}
```

gives:

```
> go mod tidy
> go run main.go
[stdout]
{"x":5}
```

But then maybe I misunderstood what you are trying to do here. 

Is there are reason you are looking to invoke `cmd/cue` in this way?

**Reply by uris77** [reply]
Thank you, this is helpful. I'll try this approach too. The reason I used `cmd/cue` was because it was not clear to me how to do this with cue's API. I figured I can just invoke whatever `cmd/cue` is doing.

**Reply by myitcv** [reply]
Whilst it might not currently be 100% the case, it is our goal that `cmd/cue` only uses public APIs. So anything that `cmd/cue` can do, you should also be able to do with the Go APIs, albeit probably more customised to your requirements.

**Reply by uris77** [reply]
How are tags injected using the Go API? I tried altering your example to this, but it didn't work:


```go
bps := load.Instances([]string{".",
		"--expression=exportedVar",
		"--inject title=myTitle",
	}, dirCfg)
```

**Reply by uris77** [reply]
nvm, I figured it out. Tags go in the `Config`.
